File: content/problems/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/solutions/dfs-from-each-source/Solution.java
class Solution {
    public List<List<Integer>> getAncestors(int n,int[][] edges){
        List<Integer>[] graph=new List[n];
        List<List<Integer>> answer=new ArrayList<>();
        for(int i=0;i<n;i++){
            graph[i]=new ArrayList<>();
            answer.add(new ArrayList<>());
        }
        for(int[] edge:edges)graph[edge[0]].add(edge[1]);
        for(int source=0;source<n;source++){
            boolean[] seen=new boolean[n];
            Deque<Integer> stack=new ArrayDeque<>(graph[source]);
            while(!stack.isEmpty()){
                int node=stack.pop();
                if(seen[node])continue;
                seen[node]=true;
                answer.get(node).add(source);
                for(int next:graph[node])stack.push(next);
            }
        }
        return answer;
    }
}

File: content/problems/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/solutions/dfs-from-each-source/solution.cpp
class Solution {
    public:vector<vector<int>> getAncestors(int n,vector<vector<int>>& edges){
        vector<vector<int>> graph(n),answer(n);
        for(auto& edge:edges)graph[edge[0]].push_back(edge[1]);
        for(int source=0;source<n;++source){
            vector<bool> seen(n);
            vector<int> stack=graph[source];
            while(!stack.empty()){
                int node=stack.back();
                stack.pop_back();
                if(seen[node])continue;
                seen[node]=true;
                answer[node].push_back(source);
                for(int next:graph[node])stack.push_back(next);
            }
        }
        return answer;
    }
};
